ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/Node.hs

```diff
@@ -0,0 +1,443 @@
+    , saSeed                  :: StdGen
+      -- ^ 'randomBlockGenerationArgs' seed argument
+    , saQuota                 :: Int
+      -- ^ 'randomBlockGenerationArgs' quota value
+    , saMbTime                :: Maybe DiffTime
+      -- ^ 'LimitsAndTimeouts' argument
+    , saRelays                :: [RelayAccessPoint]
+      -- ^ 'Interfaces' relays auxiliary value
+    , saRng                   :: StdGen
+      -- ^ 'Interfaces' 'iRng' value
+    , saDomainMap             :: Map Domain [IP]
+      -- ^ 'Interfaces' 'iDomainMap' value
+    , saAddr                  :: NtNAddr
+      -- ^ 'Arguments' 'aIPAddress' value
+    , saLocalRootPeers        :: [(Int, Map RelayAccessPoint PeerAdvertise)]
+      -- ^ 'Arguments' 'LocalRootPeers' values
+    , saLocalSelectionTargets :: PeerSelectionTargets
+      -- ^ 'Arguments' 'aLocalSelectionTargets' value
+    , saDNSTimeoutSctipt      :: Script DNSTimeout
+      -- ^ 'Arguments' 'aDNSTimeoutScript' value
+    , saDNSLookupDelayScript  :: Script DNSLookupDelay
+      -- ^ 'Arguments' 'aDNSLookupDelayScript' value
+    }
+    deriving (Show)
+
+data Command = JoinNetwork DiffTime
+             | Kill DiffTime
+             | Reconfigure DiffTime
+                           [(Int, Map RelayAccessPoint PeerAdvertise)]
+  deriving (Show, Eq)
+
+-- | Generate DNS table
+genDomainMap :: [RelayAccessPoint] -> Gen (Map Domain [IP])
+genDomainMap raps = do
+  let domains = [ d | RelayAccessDomain d _ <- raps ]
+  m <- mapM (\d -> do
+    size <- chooseInt (1, 5)
+    ips <- vectorOf size genIP
+    return (d, ips)) domains
+
+  return (Map.fromList m)
+
+  where
+    genIP :: Gen IP
+    genIP =
+      let genIPv4 = IPv4 . toIPv4 <$> replicateM 4 (choose (0,255))
+          genIPv6 = IPv6 . toIPv6 <$> replicateM 8 (choose (0,0xffff))
+       in oneof [genIPv4, genIPv6]
+
+genCommand :: [(Int, Map RelayAccessPoint PeerAdvertise)] -> Gen [Command]
+genCommand localRoots = sized $ \size -> do
+  commands <- vectorOf size (oneof [ JoinNetwork <$> delay
+                                   , Reconfigure <$> delay <*> sublistOf localRoots
+                                   , Kill <$> delay
+                                   ])
+  return (fixupCommands commands)
+  where
+    delay = frequency [(1, pure 3600), (3, (/ 10) <$> genDelayWithPrecision 100)]
+    fixupCommands :: [Command] -> [Command]
+    fixupCommands [] = []
+    fixupCommands (h:t) = go h t
+      where
+        go :: Command -> [Command] -> [Command]
+        go _ [] = []
+        go prev (cmd:cmds) =
+          case (prev, cmd) of
+            (JoinNetwork _, JoinNetwork dt) ->
+              JoinNetwork dt : go cmd cmds
+            (Kill _, Kill dt) ->
+              Kill dt : go cmd cmds
+            (Kill dt, Reconfigure _ _) ->
+              Kill dt : go prev cmds
+            (Reconfigure dt lrp, JoinNetwork _) ->
+              Reconfigure dt lrp : go prev cmds
+            _ -> prev : go cmd cmds
```

Comment:
The drawback of this approach is that the generated list of commands usually will be much shorter than `size`; this can be addressed by rescaling the generator to something larger than the default `100`.

Reply:
That's true! Should I rescale it in the generator?

Reply:
The question is how much...

Reply:
It is going to depend on how much `fixupCommands` shrinks the initial list, I'll investigate this.

Reply:
From what I have seen it does not reduce the list that much, might be proven wrong when deploying more property tests and checking coverage. But my idea is that we do not want much variability in the commands, just enough to test all scenarios.

Reply:
ok, let's see how it goes.